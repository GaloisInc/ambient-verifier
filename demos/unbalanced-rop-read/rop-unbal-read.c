#include <unistd.h>
#include <sys/syscall.h>
#include <sys/types.h>

const char *bin_sh = "/bin/sh";

void artificial_gadget_1() {
  asm("pop %rsi; pop %rdx; ret");
}

void artificial_gadget_2() {
  asm("pop %rax; pop %rdi; ret");
}

void artificial_gadget_3() {
  asm("syscall; ret");
}

void exploitable_read(int fd) {
    char name[64];
    syscall(SYS_read, fd, name, 500);
    return;
}

void read_if_root(int fd) {
  if (isatty(fd)) {
    // `fd` cannot be a TTY
    return;
  }
  if (getuid() == 0) {
    // Must be root
    exploitable_read(fd);
  }
}

int main(void) {
    int fd = 0;
    if (getppid() == 1) {
      // Must be running under init
      read_if_root(fd);
    }
    return 0;
}

void _start(void) {
    main();
}
