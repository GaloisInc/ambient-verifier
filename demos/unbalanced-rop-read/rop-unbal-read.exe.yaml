statechart:
  name: ROPViaRead
  root state:
    name: Root
    initial: NormalExecution
    states:
      - name: NormalExecution
        description: |
          Execution before the Weird Machine starts
        transitions:
          - target: WMEntered
            event: EnterWeirdMachine(0x401004)
      - name: WMEntered
        description: |
          The Weird Machine has been entered by clobbering the return address
        transitions:
          # Uncomment this transition to observe a policy violation failure
          #
          # - target: PolicyViolation
          #   event: IssuesSyscall(mkdir)
          - target: DidExec
            event: IssuesSyscall(execve)
      - name: PolicyViolation
        description: |
          Something in the payload caused the intended policy to be violated
      - name: DidExec
        description: |
          The ROP attack has completed
        type: final
