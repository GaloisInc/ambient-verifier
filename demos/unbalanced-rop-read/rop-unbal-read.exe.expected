successful: 0
failed: 0
fsRoot: tests/filesystems/rop-unbal-read
